zmq_addon.hpp

```diff
@@ -408,190 +408,115 @@ class multipart_t
-
-        multipart_t received(input);
-        assert(received.size() == 6);
-
-        str = received.popstr();
-        assert(received.size() == 5);
-        assert(str == "One");
-
-        str = received.popstr();
-        assert(received.size() == 4);
-        assert(str == "Two");
-
-        num = received.poptyp<float>();
-        assert(received.size() == 3);
-        assert(num == 49.0f);
-
-        str = received.popstr();
-        assert(received.size() == 2);
-        assert(str == "Fifty");
-
-        str = received.popstr();
-        assert(received.size() == 1);
-        assert(str == "Ninety-nine");
-
-        str = received.popstr();
-        assert(received.empty());
-        assert(str == "One-hundred");
-
-        return 0;
-    }
-
 private:
     // Disable implicit copying (moving is more efficient)
     multipart_t(const multipart_t& other) ZMQ_DELETED_FUNCTION;
     void operator=(const multipart_t& other) ZMQ_DELETED_FUNCTION;
 };  // class multipart_t
 
-#endif  // ZMQ_HAS_RVALUE_REFS
-
 inline std::ostream& operator<<(std::ostream& os, const multipart_t& msg)
 {
     return os << msg.str();
 }
 
+#endif  // ZMQ_HAS_RVALUE_REFS
+
+#if defined(ZMQ_BUILD_DRAFT_API) && defined(ZMQ_CPP11) && defined(ZMQ_HAVE_POLLER)
+    class active_poller_t
+    {
+    public:
+        active_poller_t () = default;
+        ~active_poller_t () = default;
+
+        active_poller_t(const active_poller_t&) = delete;
+        active_poller_t &operator=(const active_poller_t&) = delete;
+
+        active_poller_t(active_poller_t&& src) = default;
+        active_poller_t &operator=(active_poller_t&& src) = default;
+
+        using handler_t = std::function<void(short)>;
+
+        void add (zmq::socket_t &socket, short events, handler_t handler)
+        {
+            auto it = decltype (handlers)::iterator {};
+            auto inserted = bool {};
+            std::tie(it, inserted) = handlers.emplace (static_cast<void*>(socket), std::make_shared<handler_t> (std::move (handler)));
+            try 
+            {
+                base_poller.add (socket, events, inserted && *(it->second) ? it->second.get() : nullptr);
+                need_rebuild |= inserted;
+            }
+            catch (const zmq::error_t&) 
+            {
+                // rollback
+                if (inserted)
+                {
+                    handlers.erase (static_cast<void*>(socket));
+                }
+                throw;
+            }
+        }
+
+        void remove (zmq::socket_t &socket)
+        {
+            base_poller.remove (socket);
+            handlers.erase (static_cast<void*>(socket));
+            need_rebuild = true;
+        }
+
+        void modify (zmq::socket_t &socket, short events)
+        {
+            base_poller.modify (socket, events);
+        }
+
+        int wait (std::chrono::milliseconds timeout)
+        {
+            if (need_rebuild) {
+                poller_events.resize (handlers.size ());
```

Comment:
This is wrong, sorry. resize does not clear items already in the vector so you might pass items from previous run to wait. Depending on this kind of stuff is unreliable and asking for trouble.

Reply:
That's why I changed this. It is not necessary to clear the items. zmq_poller_wait_all even clears all unused items. But this is unnecessary too. zmq_poller_wait_all returns the number of valid items, and we only access those.

If this were different, you sould need to clear all items on every call to wait, regardless of the value of need_rebuild. But this would be quite inefficient if you had a large pollset.

Reply:
I get a point about performance but my issue was that we rely on internal implementation of libzmq to clear them (set all members) for us. How about if `libzmq` decides to set only polleritem partially depends on the case e.g. ZMQ_POLLERR etc.. That was my original reason behind clear().

Reply:
I think the previous version did not clear the members of the event struct either.

We shouldn't rely on implementation details of course, but I don't think we do. We should rely on the specification of the API, which should specify that the indicated number of entries are properly set to be well usable IMO. However, I fear the zmq_poller_* functions are not properly documented yet. I will put this forward on the libzmq side.

Reply:
Sounds good.